File: common/core/src/main/java/hu/bme/mit/theta/core/type/MultiaryExpr.java
/*
 *  Copyright 2017 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.collect.ImmutableList.toImmutableList;

import java.util.List;

import com.google.common.collect.ImmutableList;

import hu.bme.mit.theta.common.Utils;
import hu.bme.mit.theta.core.utils.TypeUtils;

public abstract class MultiaryExpr<OpType extends Type, ExprType extends Type> implements Expr<ExprType> {

	private final List<Expr<OpType>> ops;

	private volatile int hashCode = 0;

	protected MultiaryExpr(final Iterable<? extends Expr<OpType>> ops) {
		checkNotNull(ops);
		this.ops = ImmutableList.copyOf(ops);
	}

	@Override
	public final int getArity() {
		return ops.size();
	}

	@Override
	public final List<Expr<OpType>> getOps() {
		return ops;
	}

	@Override
	public final MultiaryExpr<OpType, ExprType> withOps(final List<? extends Expr<?>> ops) {
		checkNotNull(ops);
		if (ops.isEmpty()) {
			return with(ImmutableList.of());
		} else {
			final OpType opType = getOps().get(0).getType();
			final List<Expr<OpType>> newOps = ops.stream().map(op -> TypeUtils.cast(op, opType))
					.collect(toImmutableList());
			return with(newOps);
		}
	}

	@Override
	public final int hashCode() {
		int result = hashCode;
		if (result == 0) {
			result = getHashSeed();
			result = 31 * result + getOps().hashCode();
			hashCode = result;
		}
		return result;
	}

	@Override
	public final String toString() {
		return Utils.lispStringBuilder(getOperatorLabel()).addAll(ops).toString();
	}

	public abstract MultiaryExpr<OpType, ExprType> with(final Iterable<? extends Expr<OpType>> ops);

	protected abstract int getHashSeed();

	public abstract String getOperatorLabel();

}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/MultiaryExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type

import hu.bme.mit.theta.common.Utils
import hu.bme.mit.theta.core.utils.TypeUtils
import kotlinx.serialization.Serializable

/**
 * Base class for expressions with multiple operands of the same type.
 *
 * @param OpType The type of the operands
 * @param ExprType The type of the expression
 */
@Serializable
abstract class MultiaryExpr<OpType : Type, ExprType : Type> : Expr<ExprType> {

    abstract override val ops: List<Expr<OpType>>

    abstract val operatorLabel: String

    override fun withOps(ops: List<Expr<*>>): MultiaryExpr<OpType, ExprType> {
        if (ops.isEmpty()) {
            return with(listOf())
        } else {
            val opType: OpType = this.ops[0].type
            val newOps: List<Expr<OpType>> = ops.map { op: Expr<*> -> TypeUtils.cast(op, opType) }
            return with(newOps)
        }
    }

    override fun toString(): String = Utils.lispStringBuilder(operatorLabel).body().addAll(ops).toString()

    open fun with(ops: Iterable<Expr<OpType>>): MultiaryExpr<OpType, ExprType> =
        if (ops.toList() == this.ops) {
            this
        } else {
            new(ops.toList())
        }

    protected abstract fun new(ops: List<Expr<OpType>>): MultiaryExpr<OpType, ExprType>
}
